File: src/nim/bubix1/nativemenu.m
/*
	BubiX1turboZ - the whole menu bar, built directly as NSMenu/NSMenuItem
	objects.

	The original eX1turboZ menu bar this app mirrors is two levels deep
	(Control > Save State > State 0, Device > Sound > PSG, ...), which is
	ordinary AppKit but was beyond the GUI library this port started on. The
	bar is now built here in full, application menu included, so there is one
	answer to "where does this menu item come from".

	SDL creates its own menu bar when it registers the application: an app
	menu whose items are English literals, plus Window and View menus this
	app has no use for. Installing a bar here replaces all of that, so the
	six standard application-menu items are created with translated titles
	instead of being renamed afterwards, and nothing sits on the bar that
	this app does not offer.

	Shutdown is not one of the reasons. Every way of quitting is graceful
	whichever menu bar is installed - see the header of bubix1turboz.nim
	for why, and for why the delegate is not where to hook it.

	Items are addressed by an integer tag rather than by pointer: every item
	shares one target/action pair here, which forwards the tag to a single C
	callback. That gives the Nim side one dispatch point to map tags to
	closures, instead of one closure per item - and sidesteps the trap
	documented in DevelopmentPlan phase 7, where several closures created in
	one Nim for-loop all share a single binding of the loop variable.
*/

#import <Cocoa/Cocoa.h>

typedef void (*bx1_nmenu_action_fn)(int tag);

static bx1_nmenu_action_fn action_fn = NULL;

@interface BX1MenuTarget : NSObject
- (void)menuAction:(id)sender;
@end

@implementation BX1MenuTarget
- (void)menuAction:(id)sender
{
	if (action_fn != NULL) {
		action_fn((int)[(NSMenuItem *)sender tag]);
	}
}
@end

// NSMenuItem holds its target weakly, so this must stay alive for the
// process's lifetime or every menu click becomes a silent no-op (or worse,
// a message to a freed object). Deliberately allocated once and never
// released.
static BX1MenuTarget *target = nil;
// tag -> NSMenuItem, so callers can re-title / check / enable an item later
// without holding pointers across the FFI boundary.
static NSMutableDictionary *items_by_tag = nil;

/*
	Menu titles can carry bytes that are not UTF-8, and +stringWithUTF8String:
	returns nil for those - which would then reach -setTitle: and raise
	NSInvalidArgumentException. This is not hypothetical for X1 software: the
	core stores a D88's 17-byte disk name verbatim on non-Windows builds (its
	MultiByteToWideChar conversion is inside #ifdef _UNICODE, i.e. Win32
	only), so a Japanese multi-disk image hands us raw Shift-JIS, as do
	Shift-JIS filenames coming out of bsdtar. Try UTF-8, then Shift-JIS, and
	fall back to a lossy read so a title is always produced.

	Not static: filedialog.m's disk picker shows the same D88 names and has
	to decode them the same way. One decoder, one place to fix.
*/
NSString *bx1_ns_string(const char *bytes)
{
	NSString *s;

	if (bytes == NULL) {
		return @"";
	}
	if ((s = [NSString stringWithUTF8String:bytes]) != nil) {
		return s;
	}
	s = [NSString stringWithCString:bytes encoding:NSShiftJISStringEncoding];
	if (s != nil) {
		return s;
	}
	s = [[[NSString alloc] initWithBytes:bytes length:strlen(bytes)
		encoding:NSISOLatin1StringEncoding] autorelease];
	return s != nil ? s : @"";
}

static void ensure_globals(void)
{
	if (target == nil) {
		target = [[BX1MenuTarget alloc] init];
	}
	if (items_by_tag == nil) {
		items_by_tag = [[NSMutableDictionary alloc] init];
	}
}

void bx1_nmenu_set_action(bx1_nmenu_action_fn fn)
{
	action_fn = fn;
}

/*
	Every menu created here turns off Cocoa's automatic item validation.

	With it on (the default), AppKit re-derives every item's enabled state
	from its target/action pair right before the menu is displayed, silently
	discarding any manual setEnabled: call on an item that has one - which
	every item here does. On its own that would only make
	bx1_nmenu_set_enabled a no-op, but there is a worse failure mode: a
	modal panel that leaves NSApp's modal-session bookkeeping in a broken
	state (logged by AppKit as "modalSession has been exited prematurely")
	makes autoenablesItems treat every item outside that phantom modal
	window as invalid and disable it - which reads as "the entire menu bar
	is now dead" after using any Open/Save dialog. Turning it off removes
	the dependency on that bookkeeping entirely, and nothing in this app
	relies on Cocoa's automatic validation. See filedialog.m, which runs its
	panels app-modal for the other half of the same problem.
*/
static NSMenu *make_menu(const char *title)
{
	NSMenu *menu = [[NSMenu alloc] initWithTitle:bx1_ns_string(title)];
	[menu setAutoenablesItems:NO];
	return menu;
}

/*
	Replaces whatever menu bar the application has with an empty one, and
	returns the application menu at the head of it - the menu macOS draws
	under the app's name, which the caller fills in with About, Hide, Quit
	and the rest.

	The application menu is simply the first item of the bar; its title is
	ignored for display (macOS substitutes the bundle name), but is set
	anyway so the item is not anonymous in the debugger.
*/
void *bx1_nmenu_install_menubar(const char *app_name)
{
	NSMenu *bar;
	NSMenu *app_menu;
	NSMenuItem *app_item;

	ensure_globals();
	if (NSApp == nil) {
		return NULL;
	}
	bar = [[NSMenu alloc] initWithTitle:@""];
	[bar setAutoenablesItems:NO];
	app_menu = make_menu(app_name);
	app_item = [[NSMenuItem alloc] initWithTitle:bx1_ns_string(app_name)
		action:NULL keyEquivalent:@""];
	[app_item setSubmenu:app_menu];
	[bar addItem:app_item];
	[NSApp setMainMenu:bar];
	return app_menu;
}

/*
	Appends one of the application menu's standard items - the ones whose
	behaviour belongs to AppKit rather than to this app, so they carry an
	AppKit selector instead of a tag.

	`which`: 0 = Services, 1 = Hide, 2 = Hide Others, 3 = Show All. Services
	is a submenu AppKit populates itself once it has been handed to NSApp.
	The others go to nil, i.e. down the responder chain to NSApp.
*/
void bx1_nmenu_add_standard_item(void *menu, const char *title, int which,
	const char *key, int mods)
{
	NSMenuItem *item;
	SEL action = NULL;

	ensure_globals();
	if (menu == NULL) {
		return;
	}
	switch (which) {
	case 1: action = @selector(hide:); break;
	case 2: action = @selector(hideOtherApplications:); break;
	case 3: action = @selector(unhideAllApplications:); break;
	default: break;
	}
	item = [[NSMenuItem alloc] initWithTitle:bx1_ns_string(title)
		action:action keyEquivalent:@""];
	if (which == 0) {
		NSMenu *services = [[NSMenu alloc] initWithTitle:@""];

		// Left on automatic validation, unlike every other menu here: its
		// items are AppKit's and only AppKit knows which of them apply to
		// what is currently selected.
		[item setSubmenu:services];
		[NSApp setServicesMenu:services];
	} else {
		// The menus here have automatic validation off, so an item that
		// AppKit would normally enable for itself has to say so.
		[item setEnabled:YES];
	}
	if (key != NULL && key[0] != '\0') {
		NSEventModifierFlags mask = 0;
		if (mods & 1) mask |= NSEventModifierFlagCommand;
		if (mods & 2) mask |= NSEventModifierFlagShift;
		if (mods & 4) mask |= NSEventModifierFlagOption;
		if (mods & 8) mask |= NSEventModifierFlagControl;
		[item setKeyEquivalent:[NSString stringWithUTF8String:key]];
		[item setKeyEquivalentModifierMask:mask];
	}
	[(NSMenu *)menu addItem:item];
}

void *bx1_nmenu_add_toplevel(const char *title)
{
	NSMenu *bar;
	NSMenuItem *item;
	NSMenu *menu;

	ensure_globals();
	if (NSApp == nil || (bar = [NSApp mainMenu]) == nil) {
		return NULL;
	}
	menu = make_menu(title);
	item = [[NSMenuItem alloc] initWithTitle:bx1_ns_string(title)
		action:NULL keyEquivalent:@""];
	[item setSubmenu:menu];
	[bar addItem:item];
	return menu;
}

void *bx1_nmenu_add_submenu(void *parent, const char *title)
{
	NSMenu *menu;
	NSMenuItem *item;

	ensure_globals();
	if (parent == NULL) {
		return NULL;
	}
	menu = make_menu(title);
	item = [[NSMenuItem alloc] initWithTitle:bx1_ns_string(title)
		action:NULL keyEquivalent:@""];
	[item setSubmenu:menu];
	[(NSMenu *)parent addItem:item];
	return menu;
}

/*
	`key`: a single character for the Cmd-key equivalent, or NULL/"" for
	none. `mods` is a bitmask: 1 = Command, 2 = Shift, 4 = Option,
	8 = Control. Command is not implied - pass it explicitly - so an item
	can carry a bare function-key equivalent if one is ever wanted.
*/
void bx1_nmenu_add_item(void *menu, const char *title, int tag, const char *key, int mods)
{
	NSMenuItem *item;

	ensure_globals();
	if (menu == NULL) {
		return;
	}
	item = [[NSMenuItem alloc] initWithTitle:bx1_ns_string(title)
		action:@selector(menuAction:) keyEquivalent:@""];
	[item setTarget:target];
	[item setTag:tag];
	if (key != NULL && key[0] != '\0') {
		NSEventModifierFlags mask = 0;
		if (mods & 1) mask |= NSEventModifierFlagCommand;
		if (mods & 2) mask |= NSEventModifierFlagShift;
		if (mods & 4) mask |= NSEventModifierFlagOption;
		if (mods & 8) mask |= NSEventModifierFlagControl;
		[item setKeyEquivalent:[NSString stringWithUTF8String:key]];
		[item setKeyEquivalentModifierMask:mask];
	}
	[(NSMenu *)menu addItem:item];
	[items_by_tag setObject:item forKey:[NSNumber numberWithInt:tag]];
}

// Registered under a tag like any other item so it can be hidden, which
// matters for a separator that only makes sense when the optional section
// below it is showing.
void bx1_nmenu_add_separator(void *menu, int tag)
{
	NSMenuItem *item;

	ensure_globals();
	if (menu == NULL) {
		return;
	}
	item = [NSMenuItem separatorItem];
	[(NSMenu *)menu addItem:item];
	if (tag != 0) {
		[items_by_tag setObject:item forKey:[NSNumber numberWithInt:tag]];
	}
}

static NSMenuItem *item_for_tag(int tag)
{
	if (items_by_tag == nil) {
		return nil;
	}
	return [items_by_tag objectForKey:[NSNumber numberWithInt:tag]];
}

void bx1_nmenu_set_checked(int tag, int checked)
{
	[item_for_tag(tag) setState:(checked ? NSControlStateValueOn : NSControlStateValueOff)];
}

int bx1_nmenu_get_checked(int tag)
{
	NSMenuItem *item = item_for_tag(tag);
	return (item != nil && [item state] == NSControlStateValueOn) ? 1 : 0;
}

void bx1_nmenu_set_enabled(int tag, int enabled)
{
	[item_for_tag(tag) setEnabled:(enabled != 0)];
}

// Hiding rather than disabling is how the original renders its optional
// sections: the D88 bank list and the recent-file entries simply are not
// there when they do not apply (winmain.cpp rebuilds the menu tail each
// time). AppKit collapses hidden items out of the layout, giving the same
// result without rebuilding anything.
void bx1_nmenu_set_hidden(int tag, int hidden)
{
	[item_for_tag(tag) setHidden:(hidden != 0)];
}

void bx1_nmenu_set_item_title(int tag, const char *title)
{
	[item_for_tag(tag) setTitle:bx1_ns_string(title)];
}

File: src/nim/bubix1/nativemenu.nim
## Menu bar construction on top of AppKit directly. See nativemenu.m.
##
## Usage: `installMenuBar` once the application object exists (SDL creates
## it), then build the menus. Actions are ordinary Nim closures:
##
## ```nim
## let appMenu = installMenuBar("BubiX1turboZ")
## appMenu.addItem("About BubiX1turboZ", proc () = showAbout())
## let control = addMenu("Control")
## control.addItem("Reset", proc () = bx1Reset(h))
## let save = control.addSubmenu("Save State")
## for i in 0 .. 9:
##   save.addItem("State " & $i, makeSaveAction(i))
## ```
##
## Every item is registered with an integer tag and one shared Objective-C
## target; this module keeps the tag -> closure table. That means a `for`
## loop can add items directly without the `makeXxxAction(i)` dance
## DevelopmentPlan phase 7 documents, since the closure is stored per tag
## rather than captured per item - but note the loop variable still has to
## be bound freshly if the closure reads it, so the pattern above stays the
## safe habit.

import std/tables

{.compile: "nativemenu.m".}
{.passL: "-framework Cocoa".}

type
  Menu* = object
    ## An NSMenu. Zero value (`handle == nil`) is a valid no-op menu, which
    ## is what every operation degrades to if the menu bar is not up yet.
    handle: pointer

  MenuAction* = proc () {.closure.}

  MenuItemRef* = object
    ## A handle for changing an item after creation (check state, enabled
    ## state, title). Cheap to copy.
    tag*: cint

proc bx1NmenuSetAction(fn: proc (tag: cint) {.cdecl.})
  {.importc: "bx1_nmenu_set_action", cdecl.}
proc bx1NmenuInstallMenubar(appName: cstring): pointer
  {.importc: "bx1_nmenu_install_menubar", cdecl.}
proc bx1NmenuAddStandardItem(menu: pointer, title: cstring, which: cint,
                             key: cstring, mods: cint)
  {.importc: "bx1_nmenu_add_standard_item", cdecl.}
proc bx1NmenuAddToplevel(title: cstring): pointer
  {.importc: "bx1_nmenu_add_toplevel", cdecl.}
proc bx1NmenuAddSubmenu(parent: pointer, title: cstring): pointer
  {.importc: "bx1_nmenu_add_submenu", cdecl.}
proc bx1NmenuAddItem(menu: pointer, title: cstring, tag: cint, key: cstring, mods: cint)
  {.importc: "bx1_nmenu_add_item", cdecl.}
proc bx1NmenuAddSeparator(menu: pointer, tag: cint)
  {.importc: "bx1_nmenu_add_separator", cdecl.}
proc bx1NmenuSetHidden(tag: cint, hidden: cint) {.importc: "bx1_nmenu_set_hidden", cdecl.}
proc bx1NmenuSetChecked(tag: cint, checked: cint) {.importc: "bx1_nmenu_set_checked", cdecl.}
proc bx1NmenuGetChecked(tag: cint): cint {.importc: "bx1_nmenu_get_checked", cdecl.}
proc bx1NmenuSetEnabled(tag: cint, enabled: cint) {.importc: "bx1_nmenu_set_enabled", cdecl.}
proc bx1NmenuSetItemTitle(tag: cint, title: cstring)
  {.importc: "bx1_nmenu_set_item_title", cdecl.}

const
  ModCommand* = 1'i32
  ModShift* = 2'i32
  ModOption* = 4'i32
  ModControl* = 8'i32

# Module-level rather than passed through AppKit's `representedObject`: a
# Nim closure is a two-word value (proc + environment) that cannot be
# round-tripped through a single Objective-C pointer without manual GC
# rooting, and the table keeps the environment reachable for free.
var actions: Table[cint, MenuAction]
var nextTag: cint = 1
var installed = false

proc dispatch(tag: cint) {.cdecl.} =
  ## The single entry point every menu click arrives through.
  let a = actions.getOrDefault(tag)
  if a != nil:
    a()

proc ensureInstalled() =
  if not installed:
    bx1NmenuSetAction(dispatch)
    installed = true

type StandardItem* = enum
  ## The application-menu items whose behaviour is AppKit's own.
  siServices, siHide, siHideOthers, siShowAll

proc installMenuBar*(appName: string): Menu =
  ## Replaces the application's menu bar with an empty one and returns the
  ## application menu at the head of it. Everything else is appended after
  ## that menu, so this has to come first.
  ensureInstalled()
  Menu(handle: bx1NmenuInstallMenubar(appName.cstring))

proc addStandardItem*(menu: Menu, title: string, which: StandardItem,
                      key = "", mods = ModCommand) =
  ## Appends an item that AppKit acts on itself (Services, Hide, Show All).
  ensureInstalled()
  if menu.handle != nil:
    bx1NmenuAddStandardItem(menu.handle, title.cstring, which.cint,
                            key.cstring, mods.cint)

proc addMenu*(title: string): Menu =
  ## Appends a new top-level menu to the menu bar, after the application
  ## menu `installMenuBar` put there.
  ensureInstalled()
  Menu(handle: bx1NmenuAddToplevel(title.cstring))

proc addSubmenu*(parent: Menu, title: string): Menu =
  ensureInstalled()
  if parent.handle == nil:
    return Menu(handle: nil)
  Menu(handle: bx1NmenuAddSubmenu(parent.handle, title.cstring))

proc addItem*(menu: Menu, title: string, action: MenuAction = nil,
              key = "", mods = ModCommand): MenuItemRef {.discardable.} =
  ## `key` is a single character for a Cmd-key equivalent, or "" for none.
  ## `mods` is ignored when `key` is empty.
  ensureInstalled()
  result = MenuItemRef(tag: nextTag)
  inc nextTag
  if action != nil:
    actions[result.tag] = action
  if menu.handle != nil:
    bx1NmenuAddItem(menu.handle, title.cstring, result.tag, key.cstring, mods.cint)

proc setAction*(item: MenuItemRef, action: MenuAction) =
  ## Replaces (or supplies) an item's action after creation. Needed for a
  ## self-toggling check item, whose closure has to refer to the item it is
  ## attached to - which does not exist yet while `addItem` is running.
  actions[item.tag] = action

proc addSeparator*(menu: Menu): MenuItemRef {.discardable.} =
  ## The returned handle only matters for a separator that has to be hidden
  ## along with an optional section below it.
  ensureInstalled()
  result = MenuItemRef(tag: nextTag)
  inc nextTag
  if menu.handle != nil:
    bx1NmenuAddSeparator(menu.handle, result.tag)

proc `hidden=`*(item: MenuItemRef, value: bool) =
  ## Hidden items collapse out of the menu entirely, which is how the
  ## original app presents its optional sections (the D88 bank list and the
  ## recent-file entries are simply absent when they do not apply).
  bx1NmenuSetHidden(item.tag, value.cint)

proc `checked=`*(item: MenuItemRef, value: bool) =
  bx1NmenuSetChecked(item.tag, value.cint)

proc checked*(item: MenuItemRef): bool =
  bx1NmenuGetChecked(item.tag) != 0

proc `enabled=`*(item: MenuItemRef, value: bool) =
  bx1NmenuSetEnabled(item.tag, value.cint)

proc `title=`*(item: MenuItemRef, value: string) =
  bx1NmenuSetItemTitle(item.tag, value.cstring)
